pkg/cli/tail.go

```diff
@@ -222,24 +275,46 @@ func (c *TailCommand) Run(ctx context.Context, args []string) error {
 	return nil
 }
 
-func (c *TailCommand) tail(ctx context.Context) ([]*loggingpb.LogEntry, error) {
-	var p logPuller
-	if c.testPuller != nil {
-		p = c.testPuller
-	} else {
-		logClient, err := logging.NewClient(ctx)
+func (c *TailCommand) stream(ctx context.Context, extra []validation.Validator, puller logPuller) error {
+	// logCh is used to receive logEntriesfrom StreamPull
+	logCh := make(chan *loggingpb.LogEntry)
+
+	var failCount, totalCount int
+
+	var perr error
+	go func() {
+		if err := puller.StreamPull(ctx, c.queryFilter(), logCh); err != nil {
+			perr = fmt.Errorf("StreamPull failed: %w", err)
+		}
+		close(logCh)
+	}()
+
+	for l := range logCh {
+		totalCount++
+		js, err := protojson.Marshal(l)
 		if err != nil {
-			return nil, fmt.Errorf("failed to create logging client: %w", err)
+			c.Errf("failed to marshal log to json (InsertId: %q): %w", l.InsertId, err)
+			continue
+		}
+		c.Outf(stripSpaces(string(js)))
+		if c.flagValidate {
+			if err := validation.Validate(string(js), extra...); err != nil {
+				failCount++
+				c.Errf("failed to validate log (InsertId: %q): %w\n", l.InsertId, err)
+			} else {
+				c.Outf("Successfully validated log (InsertId: %q)\n", l.InsertId)
+			}
 		}
-		p = cloudlogging.NewPuller(ctx, logClient, c.flagScope)
 	}
 
-	ls, err := p.Pull(ctx, c.queryFilter(), c.flagMaxNum)
-	if err != nil {
-		return nil, fmt.Errorf("failed to pull logs: %w", err)
+	if c.flagValidate {
+		c.Outf("Validation failed for %d logs (out of %d)", failCount, totalCount)
 	}
 
-	return ls, nil
+	if perr != nil {
+		return perr
+	}
+	return fmt.Errorf("stream tail validate cancelled: %w", ctx.Err())
```

Comment:
We probably shouldn't return an error here. It's expected if context is cancelled (triggered by the user or timeout), we end the stream. Probably use a debug log instead.

Reply:
Now use logger.DebugContext() to log this.